src/charm.py

```diff
@@ -3,115 +3,267 @@
+            try:
+                self._stored.runner_bin_url = runner_manager.get_latest_runner_bin_url()
+                runner_manager.update_runner_bin(self._stored.runner_bin_url)
+            except Exception as e:
+                logger.exception("Failed to update runner binary")
+                self.unit.status = BlockedStatus(f"Failed to update runner binary: {e}")
+                return
+            self.unit.status = MaintenanceStatus("Starting runners")
+            try:
+                runner_manager.reconcile(self.config["quantity"])
+            except RunnerError as e:
+                logger.exception("Failed to start runners")
+                self.unit.status = BlockedStatus(f"Failed to start runners: {e}")
+            else:
+                self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    def _on_upgrade_charm(self, event):
+        RunnerManager.install_deps()
 
     def _on_config_changed(self, event):
-        if self._stored.path:
-            if self._stored.path != self.config["path"]:
-                logger.error("Changing path not supported")
-                self.unit.status = BlockedStatus("Path can not be changed")
+        self._ensure_event_timer("update-runner-bin", self.config["update-interval"])
+        self._ensure_event_timer("reconcile-runners", self.config["reconcile-interval"])
+
+        if self.config["path"] != self._stored.path:
+            prev_runner_manager = self._get_runner_manager(path=self._stored.path)
+            if prev_runner_manager:
+                self.unit.status = MaintenanceStatus(
+                    "Removing runners from old org/repo"
+                )
+                prev_runner_manager.clear()
+            self._stored.path = self.config["path"]
+
+        runner_manager = self._get_runner_manager()
+        if runner_manager:
+            self.unit.status = ActiveStatus()
+        else:
+            self.unit.status = BlockedStatus("Missing token or org/repo path config")
+
+    def _on_update_runner_bin(self, event):
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            return
+        old_status = self.unit.status
+        self.unit.status = MaintenanceStatus("Checking for runner updates")
+        runner_bin_url = runner_manager.get_latest_runner_bin_url()
+        if runner_bin_url != self._stored.runner_bin_url:
+            self.unit.status = MaintenanceStatus("Updating runner binary")
+            try:
+                runner_manager.update_runner_bin(runner_bin_url)
+            except Exception as e:
+                logger.exception("Failed to update runner binary")
+                self.unit.status = BlockedStatus(f"Failed to update runner binary: {e}")
                 return
+            self._stored.runner_bin_url = runner_bin_url
+            # TODO: Flush existing runners? What if they're processing a job?
```

Comment:
Does this need to exist at the Juju level? I ask because a number of services in Charmed Kubeflow handle the reconciliation of similar things themselves. The equivalent here would probably be something like the operator installing an extra systemd timer that handles flushing existing runners if they're done processing everything and there's a new binary.

If this needs to exist at the Juju level (i.e. supporting `juju run-action update-runner-bin`), then the above doesn't work as well